AnnService/src/Core/SPANN/SPANNIndex.cpp

```diff
@@ -416,15 +417,20 @@ ErrorCode Index<T>::SearchIndexIterative(QueryResult &p_headQuery, QueryResult &
     resultCount = 0;
     while (continueSearch && resultCount < p_batch)
     {
+        bool oldRelaxedMono = p_extraWorkspace->m_relaxedMono;
         ErrorCode ret = SearchDiskIndexIterative(p_headQuery, p_query, p_extraWorkspace);
         bool found = (ret == ErrorCode::Success);
         if (!found && ret != ErrorCode::VectorNotFound) return ret;
         p_extraWorkspace->m_loadPosting = false;
         if (!found)
         {
+            p_headQueryResults->SetResultNum(m_options.m_headBatch);
             p_headQueryResults->Reset();
             continueSearch = m_index->SearchIndexIterativeFromNeareast(*p_headQueryResults, p_indexWorkspace, false);
             p_extraWorkspace->m_loadPosting = true;
+
+            if (!oldRelaxedMono && p_extraWorkspace->m_relaxedMono)
```

Comment:
Note that this needs test change since we can return less than the batch argument, to give client a chance to decide whether to search further before doing more IO